sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs

```diff
@@ -373,14 +373,39 @@ private async Task<IReadOnlyList<ServiceBusReceivedMessage>> ReceiveMessagesAsyn
                 // The session won't be closed in the case that MaxConcurrentCallsPerSession > 1, but with concurrency, it is not possible to guarantee ordering.
                 if (_isSessionReceiver && (!_isProcessor || SessionId != null) && messageList.Count < maxMessages)
                 {
-                    await link.DrainAsyc(cancellationToken).ConfigureAwait(false);
-
-                    // These workarounds are necessary in order to resume prefetching after the link has been drained
-                    // https://github.com/Azure/azure-amqp/issues/252#issuecomment-1942734342
-                    if (_prefetchCount > 0)
+                    try
                     {
-                        link.Settings.TotalLinkCredit = 0;
-                        link.SetTotalLinkCredit((uint)_prefetchCount, true, true);
+                        ServiceBusEventSource.Log.DrainLinkStart(Identifier);
+                        await link.DrainAsyc(cancellationToken).ConfigureAwait(false);
+                        ServiceBusEventSource.Log.DrainLinkComplete(Identifier);
+                    }
+                    catch (Exception ex)
+                    {
+                        ServiceBusEventSource.Log.DrainLinkException(Identifier, ex.ToString());
+
+                        // Drain failure that does not fault AMQP link will cause ordering to be violated. Force close
+                        // AMQP link to ensure ordering is maintained.
+                        // Detailed explanation as to why ordering is violated: https://github.com/Azure/azure-sdk-for-net/issues/47822
+                        if (_prefetchCount > 0)
+                        {
+                            try
+                            {
+                                await link.CloseAsync(CancellationToken.None).ConfigureAwait(false);
```

Comment:
```suggestion
                                ServiceBusEventSource.Log.CloseLinkStart(Identifier);
                                _connectionScope.CloseLink(link))
```

Iooks like we've got a helper method for this one.   We can remove the try/catch around it too, as there's a "SafeClose" being called in there.   No logging though, which is odd.

https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConnectionScope.cs#L1439

Maybe we can move the logging there so that we get it everywhere.